### ISSUE TITLE
Add setting to not automatically focus the Output tab for Configuration Warnings

### ISSUE THREAD

**Issue by aleksa2808** [microsoft/vscode-cpptools#7234]
Hi,
in my projects I'm using the `compile_commands.json` file generated during a build to help out with include paths. However, since this file is in the build folder, whenever I try to do a clean build (which is often) I need to delete that folder along with the `compile_commands.json` file. As soon as I do that, I get booted from current terminal tab into the `Output - C/C++ Configuration Warnings` tab which is saying that the extension can't find the `compile_commands.json` file anymore. As I recall this was previously a toast message which was much less annoying. Is there any way to disable this behavior?

EDIT: One more time when the same behavior occurs is when I remove a certain source/header file from the build, build from the terminal and then if I have that excluded file open or switch to it the editor automatically switches from the terminal to the Output tab.

**Comment by sean-mcmanus** [comment]
There isn't currently a way to disable this.

@bobbrow @Colengms Thoughts on this?

@aleksa2808 The implementation is at https://github.com/microsoft/vscode-cpptools/blob/main/Extension/src/LanguageServer/client.ts#L131 . A setting or something could be added and checked at that spot.

**Comment by aleksa2808** [comment]
If the `compile_commands.json` file utilization is mostly there to help with locating include files, I think that the red lines that show up under include directives when this file is removed/changed are conveying enough information as to make the automatic switching to the Output tab unnecessary. If that's the case, and if the automatic switching still makes sense in some scenarios, an option to toggle this behavior would be perfect.

**Comment by sean-mcmanus** [comment]
It looks like we only show the C/C++ Configuration Warnings the 1st time per running VS Code instance -- you aren't seeing this happen repeatedly with the same VS Code instance, right?

**Comment by aleksa2808** [comment]
Yes, the automatic switch does appear to only happen once for each specific VS Code instance. However, this is exacerbated by another issue I'm having with the `compile_commands.json` file integration where oftentimes I can't get the file changes to register after, say, making a new source file in a project and rebuilding it iteratively (thus adding the new file's compile line to the `compile_commands.json` file) or clean building the project (thus deleting the current and creating a new `compile_commands.json` file). The only thing that worked for me so far w.r.t. this issue has been reloading the VS Code instance. Because of this, after each rebuild, I usually end up having to reload the instance as well for the include paths to work and then on the next rebuild that modifies or regenerates the `compile_commands.json` file I get the automatic switch to the Output tab.

**Comment by bobbrow** [comment]
My thought is that we should try to fix the broader issue (needing to reload - which is a terrible experience) instead of adding a setting for this.  Adding the setting seems to be addressing a symptom instead of the root cause.

**Comment by sean-mcmanus** [comment]
@bobbrow Yeah, but practically speaking, it's a lot easier to address the symptom than the broader issue...there could be numerous reasons to Reload Window (and hard to eliminate all the reasons).

We could potentially avoid showing the channel if the current workspace folder is the same as the previously opened one (that would eliminate the need for the setting).

**Comment by aleksa2808** [comment]
Solving that issue as well would definitely be great, although I think that these two don't need to cancel each other out. I personally would turn off the automatic tab switch if given a choice and if, as stated above, I'm not overlooking something by assuming the primary focus of the integration with the `compile_commands.json` file is locating the include files.

In retrospect, I'd probably create this issue specifically asking about the option, but because I'm not so familiar with VS Code extensions I wasn't sure if this is something that this extension is doing or if the behavior is being forced by the editor itself.

Regarding the need to reload the editor, I'll continue to observe the issue and try to figure out what's going on + give feedback. Maybe there's an issue about this already, I did notice a few similar ones. Although, while a fix for that would, again, be awesome, it's not as annoying as the tab switch. The reason for that are the scenarios in which the two occur:
1. A reload to fix include paths is a quick, purposeful action that occurs when you notice the include tooltips are broken/red. It was also made much nicer when a recent change was introduced to preserve the integrated terminal state.
2. On the other hand, the auto tab switch happens even if you don't care about the include paths at the moment and are instead in the middle of a "clean the build dir and rebuild" operation. The tab switch interrupts that flow at the mid-point and forces you to manually switch back to the integrated terminal tab to continue building. That's the annoying part.

**Comment by aleksa2808** [comment]
@sean-mcmanus I agree that there are multiple valid reasons to reload the window. One of my personals is maintenance of several smaller, related projects.

As for avoiding showing the channel if the workspace is the same as the previously opened one, I think one downside of that is that it would make the occurrence of the tab switch a bit more "magic".

**Comment by sean-mcmanus** [comment]
I'd be fine with adding a setting. Not sure if Bob is opposed to that. https://github.com/microsoft/vscode-cpptools/pull/6731 shows an example of another setting that got added so it's pretty simple to do. The biggest work would probably be for our translators to translate the description string.

The reason for not having that be disabled by default is that we wanted users to be made more aware of potential configuration problems.

**Comment by bobbrow** [comment]
I would not be opposed to a setting that lets you disable focusing the output window for configuration warnings if the setting is broad and encompasses all configuration warnings (not just compile_commands.json-related warnings).  If it's too narrow a setting, then it won't be discoverable and I'd prefer not to bloat the extension with settings used by only a few people.

I still think we should spend more time understanding why squiggles don't refresh. A reload should not be required for the scenario described.  We should already be polling for changes in the compile_commands.json file in the event that the file watchers don't work.  Are we tracking this in any other issue reports?  Are we able to reproduce the problem?

**Comment by aleksa2808** [comment]
Regarding the reproducibility question, these are the relevant versions I'm currently using:  

OS: Ubuntu 18.04
VS Code: 1.54.2
cpptools: 1.3.0-insiders
RemoteSSH: 0.65.1

**Comment by sean-mcmanus** [comment]
@bobbrow Yeah, the setting would be for all configuration warnings. See https://github.com/microsoft/vscode-cpptools/issues/7237 in regards to why a reload window may be needed if a file doesn't exist when a configuration is processed.

**Comment by Colengms** [comment]
`compile_commands.json` is used for more than just gathering include paths.  There are many command line args that influence IntelliSense.  Lack of a `compile_commands.json` seems like a configuration issue that should be warned about.  Ideally, we would find a way to avoid reporting an unnecessary configuration warning, especially if this is a common scenario.  If that's not possible, and a setting to suppress surfacing the warning is used, please have it default to not suppressing the warning, as I don't expect anyone will go looking for the configuration warning output channel.

**Comment by sean-mcmanus** [comment]
@Colengms Yeah, it would default to not suppress. Also, it wouldn't "suppress" the warning, just not invoke the "show".

**Comment by aleksa2808** [comment]
My issue with this is not that warnings are being shown, it's the interrupting nature in which they are shown. If it's possible to change this to a less intrusive behavior the mentioned setting should not be needed. If a change like that is an option, maybe something like a "Can't find compile_commands.json" toast message with a "Show details" button which opens the Output tab could work?

Also, I managed to look into the problem of not being able to locate a freshly created `compile_commands.json` file without an editor reload. I found a recent change (https://github.com/microsoft/vscode-cpptools/pull/6748) which suggested that the extension really does try to find these files even if they were created after the editor was loaded, however upon debugging this a bit I found out that the path which still contained the variables that I use (`${workspaceFolder}`) was being passed to `fs.stat`. After modifying this a newly created `compile_commands.json` file was being found without a need to reload the editor! I submitted an issue (https://github.com/microsoft/vscode-cpptools/issues/7241) and a PR (https://github.com/microsoft/vscode-cpptools/pull/7242) regarding this.

**Comment by anders-wind** [comment]
I know this mentions an option for disabling on "configuration warnings", but would it also be possible to also just disable focus change on startup (or in any case). 
I don't have any warnings, but vs-code still changes focus on startup (but delayed ~1-2s). This is extremely annoying when it happens while writing a command.

**Comment by aleksa2808** [comment]
The behavior of switching to the Output tab mid-typing in the terminal was still there the last time I used the extension. It only occured the first time a warning shows after a reload, but the IMO legitimate use cases for a reload / opening a new project are common enough that they still made the issue relevant. However my workflow changed and I haven't used the extension for some time so I'm not sure if this has been fixed in the meantime.
